File: native_library/android/AdMobANE/app/src/main/java/com/tuarua/admobane/InterstitialController.kt
/*
 *  Copyright 2018 Tua Rua Ltd.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package com.tuarua.admobane

import android.os.Bundle
import com.adobe.fre.FREContext
import com.google.android.gms.ads.interstitial.InterstitialAd
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback
import com.google.ads.mediation.admob.AdMobAdapter
import com.google.android.gms.ads.*
import com.google.gson.Gson
import com.tuarua.admobane.Position.*
import com.tuarua.frekotlin.FreKotlinController

@Suppress("JoinDeclarationAndAssignment")
class InterstitialController(override var context: FREContext?,
                             private val isPersonalised: Boolean) : FreKotlinController {

    private var _adView: InterstitialAd? = null
    private var _showOnLoad = true
    private val gson = Gson()

    fun load(unitId: String, deviceList: List<String>?, targeting: Targeting?, showOnLoad: Boolean) {
        _showOnLoad = showOnLoad
        val activity = this.context?.activity ?: return
        val context = activity.applicationContext ?: return

        val requestBuilder = AdRequest.Builder()
        if (!isPersonalised) {
            val extras = Bundle()
            extras.putString("npa", "1")
            requestBuilder.addNetworkExtrasBundle(AdMobAdapter::class.java, extras)
        }

        val configBuilder = MobileAds.getRequestConfiguration().toBuilder()
        targeting?.maxAdContentRating?.let {
            configBuilder.setMaxAdContentRating(it)
        }
        targeting?.tagForChildDirectedTreatment?.let {
            configBuilder.setTagForUnderAgeOfConsent(it)
        }
        targeting?.tagForUnderAgeOfConsent?.let {
            configBuilder.setTagForUnderAgeOfConsent(it)
        }

        configBuilder.setTestDeviceIds(deviceList)

        MobileAds.setRequestConfiguration(configBuilder.build())

        InterstitialAd.load(context, unitId, requestBuilder.build(), object : InterstitialAdLoadCallback() {
            override fun onAdFailedToLoad(adError: LoadAdError) {
                _adView = null

                dispatchEvent(Constants.ON_LOAD_FAILED, gson.toJson(AdMobEvent(INTERSTITIAL.ordinal, adError.code)))
            }

            override fun onAdLoaded(interstitialAd: InterstitialAd) {
                _adView = interstitialAd
                _adView?.fullScreenContentCallback = object: FullScreenContentCallback() {
                    override fun onAdShowedFullScreenContent() {
                        dispatchEvent(Constants.ON_OPENED, gson.toJson(AdMobEvent(INTERSTITIAL.ordinal)))
                    }

                    override fun onAdDismissedFullScreenContent() {
                        _adView = null
                        dispatchEvent(Constants.ON_CLOSED, gson.toJson(AdMobEvent(INTERSTITIAL.ordinal)))
                    }
                }
                if (_showOnLoad) {
                    interstitialAd.show(activity)
                }

                dispatchEvent(Constants.ON_LOADED, gson.toJson(AdMobEvent(INTERSTITIAL.ordinal)))
            }
        })

    }

    fun show() {
        val activity = this.context?.activity ?: return
        _adView?.show(activity)
    }

    override val TAG: String?
        get() = this::class.java.canonicalName

}

File: native_library/ios/AdMobANE/AdMobANE/InterstitialController.swift
/*
 *  Copyright 2017 Tua Rua Ltd.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import UIKit
import GoogleMobileAds
import FreSwift

class InterstitialController: NSObject, FreSwiftController, GADFullScreenContentDelegate {
    static var TAG = "InterstitialController"

    internal var context: FreContextSwift!
    private var adView: GADInterstitialAd?
    private var _showOnLoad = true
    private var _airVC: UIViewController?
    private var isPersonalised = true
    convenience init(context: FreContextSwift, isPersonalised: Bool) {
        self.init()
        self.context = context
        self.isPersonalised = isPersonalised
    }
    
    func load(airVC: UIViewController, unitId: String, deviceList: [String]?,
              targeting: Targeting?, showOnLoad: Bool) {
        _airVC = airVC
        _showOnLoad = showOnLoad
        let request = GADRequest()
        if !isPersonalised {
            let extras = GADExtras()
            extras.additionalParameters = ["npa": "1"]
            request.register(extras)
        }
        
        GADMobileAds.sharedInstance().requestConfiguration.testDeviceIdentifiers = deviceList
        
        if let t = targeting {
            if let fc = t.tagForChildDirectedTreatment {
                GADMobileAds.sharedInstance().requestConfiguration.tag(forChildDirectedTreatment: fc)
            }
            if let fua = t.tagForUnderAgeOfConsent {
                GADMobileAds.sharedInstance().requestConfiguration.tagForUnderAge(ofConsent: fua)
            }
            if let mcr = t.maxAdContentRating {
                GADMobileAds.sharedInstance().requestConfiguration.maxAdContentRating = mcr
            }
            if let contentUrl = t.contentUrl {
                request.contentURL = contentUrl
            }
            
        }
        
        GADInterstitialAd.load(withAdUnitID: unitId,
                               request: request,
                               completionHandler: { [self] ad, error in
                                
                                if let error = error as NSError? {
                                    var props = [String: Any]()
                                    props["position"] = Position.interstitial.rawValue
                                    props["errorCode"] = error.code
                                    dispatchEvent(name: Constants.ON_LOAD_FAILED, value: JSON(props).description)
                                    return
                                }
                                
                                adView = ad
                                adView?.fullScreenContentDelegate = self

                                var props = [String: Any]()
                                props["position"] = Position.interstitial.rawValue
                                dispatchEvent(name: Constants.ON_LOADED, value: JSON(props).description)
                                
                                guard let av = adView, let avc = _airVC else {return}
                                if _showOnLoad {
                                    av.present(fromRootViewController: avc)
                                }
                               }
        )
    }
    
    func show() {
        guard let av = adView, let avc = _airVC else { return }
        av.present(fromRootViewController: avc)
    }
    
    func dispose() {
        guard let av = adView else {
            return
        }
        av.fullScreenContentDelegate = nil
        adView = nil
    }
    
    func adDidPresentFullScreenContent(_ ad: GADFullScreenPresentingAd) {
        var props = [String: Any]()
        props["position"] = Position.interstitial.rawValue
        dispatchEvent(name: Constants.ON_OPENED, value: JSON(props).description)
    }
    
    func adDidDismissFullScreenContent(_ ad: GADFullScreenPresentingAd) {
        var props = [String: Any]()
        props["position"] = Position.interstitial.rawValue
        dispatchEvent(name: Constants.ON_CLOSED, value: JSON(props).description)
    }
    
}
